src/eddymotion/model/utils.py

```diff
@@ -23,11 +23,245 @@
 import numpy as np
 from dipy.core.gradients import get_bval_indices
 from sklearn.cluster import KMeans
+from sklearn.gaussian_process.kernels import Kernel
+from scipy.optimize import minimize, Bounds
 
 B0_THRESHOLD = 50  # from dmriprep
 SHELL_DIFF_THRES = 20  # 150 in dmriprep
 
 
+class SphericalCovarianceKernel(Kernel):
+    """
+    Custom kernel based on spherical covariance function.
+
+    Parameters
+    ----------
+    lambda_ : float, default=1.0
+        Scale parameter for the covariance function.
+    a : float, default=1.0
+        Distance parameter where the covariance function goes to zero.
+    sigma_sq : float, default=1.0
+        Noise variance term.
+    """
+    def __init__(self, lambda_=1.0, a=1.0, sigma_sq=1.0):
+        self.lambda_ = lambda_
+        self.a = a
+        self.sigma_sq = sigma_sq
+
+    def __call__(self, theta):
+        """
+        Compute the kernel matrix.
+
+        Parameters
+        ----------
+        theta : array-like of shape (n_samples, n_samples)
+            Precomputed pairwise angles.
+
+        Returns
+        -------
+        K : array-like of shape (n_samples, n_samples)
+            Kernel matrix.
+        """
+        K = np.where(theta <= self.a, 1 - 3 * (theta / self.a) ** 2 + 2 * (theta / self.a) ** 3, 0)
```

Comment:
I'm not sure, but maybe it would be better to compute the angles between vectors here

Reply:
My understanding is that this information is known beforehand, and IMO best would be to compute it in a separate method. That would allow us to know whether the computations are OK with some simple data and without requiring a kernel instance.

If the fit/predict (or the kernel) needs to look for some given values for a given index (not sure about this), here we would only need to get the appropriate value. The [`utils.compute_angles`](https://github.com/jhlegarreta/eddymotion/pull/2/files#diff-0433c95661a0c74c5322b894590de2ae8ee0930c5c29c917f9b0f1aa2e937822R183) should do this.

I still need to see how Yibei's work completes this.

Reply:
https://github.com/nipreps/eddymotion/pull/200 is ready to be merged. Efficiency is left for when we will be dealing with real/large. For now, I prefer to keep it things more evident. So the upstream should be called here. If we need to index some value, we can index it from the matrix IMO. Resolving.